simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/HeartbeatScheduler.java

```diff
@@ -19,24 +23,45 @@
     private StationState stationState;
     private StationMessageSender stationMessageSender;
 
+    private AtomicInteger heartBeatInterval;
+    private AtomicLong timeOfLastHeartbeatSent;
+
     public HeartbeatScheduler(StationState stationState, StationMessageSender stationMessageSender) {
         this.stationState = stationState;
         this.stationMessageSender = stationMessageSender;
+        this.heartBeatInterval = new AtomicInteger();
+        this.timeOfLastHeartbeatSent = new AtomicLong(0);
     }
 
-    public void scheduleHeartbeat(int heartbeatInterval) {
+    public void updateHeartbeat(int heartbeatInterval) {
         LOGGER.debug("Scheduling heartbeat to {} sec.", heartbeatInterval);
-        heartbeatExecutor.scheduleAtFixedRate(this::sendHeartbeat, heartbeatInterval, heartbeatInterval, TimeUnit.SECONDS);
+        heartBeatInterval.set(heartbeatInterval);
+        heartbeatExecutor.scheduleAtFixedRate(
+                this::shouldSendHeartbeat,
+                heartbeatInterval,
+                max(1, heartbeatInterval / 2),
+                TimeUnit.SECONDS);
     }
 
-    private void sendHeartbeat() {
+    private void shouldSendHeartbeat() {
         try {
-            HeartbeatRequest heartbeatRequest = new HeartbeatRequest();
-            Subscriber<HeartbeatRequest, HeartbeatResponse> subscriber = (request, response) -> stationState.setCurrentTime(response.getCurrentTime());
+            long now = System.currentTimeMillis();
```

Comment:
I'd prefer to see here Java 8 time stuff
for example now.plusHours(24).isAfter(scheduledHeartbeat) || ...
and don't deal with these milliseconds explicitly